src/main/java/com/devcycle/sdk/server/common/model/DevCycleUser.java

```diff
@@ -30,81 +35,170 @@
+    private Map<String, Object> privateCustomData;
+
+    @Schema(description = "Date the user was created, Unix epoch timestamp format")
+    @JsonProperty("createdDate")
+    @JsonDeserialize(using = LongTimestampDeserializer.class)
+    @JsonFormat(shape = JsonFormat.Shape.STRING)
+    private Long createdDate;
+
+    @Schema(description = "Date the user was last seen, Unix epoch timestamp format")
+    @JsonProperty("lastSeenDate")
+    @JsonDeserialize(using = LongTimestampDeserializer.class)
+    @JsonFormat(shape = JsonFormat.Shape.STRING)
+    private Long lastSeenDate;
+
+    @Schema(description = "Platform the SDK is running on")
+    @Builder.Default
+    @JsonProperty("platform")
+    private String platform = getPlatformData().getPlatform();
+
+    @Schema(description = "Version of the platform the SDK is running on")
+    @Builder.Default
+    @JsonProperty("platformVersion")
+    private String platformVersion = getPlatformData().getPlatformVersion();
+
+    @Schema(description = "DevCycle SDK type")
+    @Builder.Default
+    @JsonProperty("sdkType")
+    private PlatformData.SdkTypeEnum sdkType = getPlatformData().getSdkType();
+
+    @Schema(description = "DevCycle SDK Version")
+    @Builder.Default
+    @JsonProperty("sdkVersion")
+    private String sdkVersion = getPlatformData().getSdkVersion();
+
+    @Schema(description = "Hostname where the SDK is running")
+    @Builder.Default
+    @JsonProperty("hostname")
+    private String hostname = getPlatformData().getHostname();
+
+    public static PlatformData getPlatformData() {
+        return PlatformData.builder().build();
+    }
+
+    static void setCustomValue(Map<String, Object> customData, String key, Value value) {
+        // Only support boolean, number, string and null types for custom data values
+        // ignore all other data
+        if (key != null && value != null) {
+            if (value.isBoolean()) {
+                customData.put(key, value.asBoolean());
+            } else if (value.isNumber()) {
+                customData.put(key, value.asDouble());
+            } else if (value.isString()) {
+                customData.put(key, value.asString());
+            } else if (value.isNull()) {
+                customData.put(key, null);
+            } else {
+                throw new IllegalArgumentException("Custom data values must be of type boolean, number, string or null");
+            }
+        }
+    }
+
+    /**
+     * Create a DevCycleUser from an EvaluationContext
+     *
+     * @param ctx A context to load a targeting key and user data from
+     * @return An initialized DevCycleUser with data from the context
+     * @throws TargetingKeyMissingError if the targeting key or user_id attribute is not set
+     */
+    public static DevCycleUser fromEvaluationContext(EvaluationContext ctx) {
+        String userId = "";
+
+        if (ctx != null && ctx.getTargetingKey() != null) {
+            userId = ctx.getTargetingKey();
```

Comment:
Do we want to account for the targetingKey being an empty string here? Same for the user ID.

Reply:
It does check for empty lower but you are right, if someone sets `targetingKey = ""` and then provides a valid `user_id` attribute it won't fall back to the user_id properly